addons/udes_stock_refactoring/models/stock_move.py

```diff
@@ -0,0 +1,355 @@
+            elif stage == "assign":
+                refactor_action = picking_type.u_post_assign_action
+            elif stage == "validate":
+                refactor_action = picking_type.u_post_validate_action
+
+            if refactor_action:
+                refactor_class = REFACTOR_REGISTRY[refactor_action]
+                return refactor_class(self.env)
+
+            return refactor_action
+        except KeyError:
+            raise ValueError(_("Refactor action '%s' could not be found." % refactor_action))
+
+    def compute_grouping_key(self):
+        """Compute the grouping key from the picking type move key format"""
+        # TODO MTC: This can be refactored and abstracted at some point with
+        # this with the equivalent for move_line_key
+        # we need to think of how we want to do it
+
+        # The environment must include {'compute_key': True}
+        # to allow the keys to be computed.
+        if not self.env.context.get("compute_key", False):
+            return
+        for move in self:
+            move_vals = {
+                field_name: move[field_name]
+                for field_name in move._fields.keys()
+                if field_name != "u_grouping_key"
+            }
+
+            format_str = move.picking_id.picking_type_id.u_move_key_format
+
+            if format_str:
+                move.u_grouping_key = format_str.format(**move_vals)
+            else:
+                move.u_grouping_key = None
+
+    u_grouping_key = fields.Char("Key", compute="compute_grouping_key")
+
+    def action_refactor(self):
+        """
+        Refactor all the moves in self. May result in the moves being changed
+        and/or their associated pickings being deleted.
+        """
+        self._action_refactor()
+        return True
+
+    def _action_refactor(self, stage=None):
+        """
+        Refactor moves in self.
+        :param stage: One of confirm|assign|done, if set, filters the moves
+            which will be refactored to only the state(s) that match:
+                - 'confirm': confirmed, waiting
+                - 'assign': assigned, partially_available
+                - 'done': done
+
+           Methods doing a refactor are expected to take a single recordset of
+           moves on which they will act, and to return the recordset of
+           equivalent moves after they have been transformed.
+           The output moves may be identical to the input, may contain none
+           of the input moves, or anywhere in between.
+           The output should contain a functionally similar set of moves.
+        """
+        if stage is not None and stage not in U_STOCK_REFACTOR_STAGES.values():
+            raise UserError(_("Unknown stage for move refactor: %s") % stage)
+        moves = self
+
+        if self.env.context.get("disable_move_refactor"):
+            return moves
+
+        refactor_moves = moves.filtered(
```

Comment:
we could do something like 

```
refactor_lam = lambda m: m.picking_type_id and m.state not in ["draft", "cancel"]
if stage is not None:
    refactor_lam =  lambda m, lam=refactor_lam: U_STOCK_REFACTOR_STAGES[m.state] == stage and lam(m)

refactor_moves = moves.filtered(refactor_lam)
```

saves a loop over moves not sure if its worth it though probably be premature optimization.

Reply:
Have added this in, every little helps plus it cuts down the code a bit. :)